### ISSUE TITLE
Variable not defined depending on conditionals

### ISSUE THREAD

**Issue by joshcltix** [PyCQA/pylint#3229]
### Is your feature request related to a problem? Please describe
Consider the following buggy code. Running it will result in an error "NameError: name 'false_var' is not defined". However, pylint 2.4.3 doesn't catch the undefined variable problem.

```
true_var = True

if true_var:
    true_var = True
else:
    false_var = False

if true_var == false_var:
    pass
```

### Describe the solution you'd like
I would like pylint to warn about the undefined variable false_var

**Comment by PCManticore** [comment]
Thanks for the report. Right now we don't have good control flow inference for catching these kind of issues, but this issue is something we definitely need to solve.

**Comment by transhapHigsn** [comment]
I have a similar issue with a method something like this:
`def func_x(param_a, param_b):
    res_1 = param_a + param_b
    if param_a and not param_b:
        res_2 = param_a
        # doing something here.
        return res_2
    elif param_b and not param_a:
        res_2 = param_b
        # doing something here.
        return res_2
    return res_2 # here res_1 was supposed to be return
`

Response from pylint on running command: `pylint file_name,py`

`
************* Module check
check.py:1:0: C0114: Missing module docstring (missing-module-docstring)
check.py:1:0: C0116: Missing function or method docstring (missing-function-docstring)
check.py:3:4: R1705: Unnecessary "elif" after "return" (no-else-return)
check.py:2:4: W0612: Unused variable 'res_1' (unused-variable)

------------------------------------------------------------------
Your code has been rated at 5.56/10 (previous run: 5.56/10, +0.00)`

**Comment by jacobtylerwalls** [comment]
If we interpret this to be about `used-before-assignment` for `false_var` (which I think is fair) then it's a duplicate of #1727 and #267.

**Comment by Pierre-Sassoulas** [comment]
Closing as duplicate of #1727
